docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+---
+title: MachinePool Machines
+authors:
+  - "@devigned"
+  - "@jont828"
+  - "@mboersma"
+reviewers:
+  -
+creation-date: 2022-02-09
+last-updated: 2022-02-09
+status: experimental
+---
+
+# MachinePool Machines
+
+## Table of Contents
+
+- [MachinePool Machines](#machinepool-machines)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story U1](#story-u1)
+      - [Story U2](#story-u2)
+      - [Story U3](#story-u3)
+    - [Requirements](#requirements)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+* **Machine**: a declarative specification for an infrastructure component that hosts a Kubernetes Node. Commonly, a Machine represents a VM instance.
+* **MachinePool**: a spec for an infrastructure component that can scale a group of Machines. A MachinePool acts similarly to a [MachineDeployment][], but leverages scaling up resources that are specific to an infrastructure provider. However, while individual Machines in a MachineDeployment are visible and can be removed, individual Machines in a MachinePool are not represented.
+* **MachinePool Machine**: a Machine that is owned by a MachinePool. It requires no special treatment in Cluster API in general; a Machine is a Machine without regard to ownership.
+
+See also the [Cluster API Book Glossary][].
+
+## Summary
+
+MachinePools should be enhanced to own Machines that represent each of its `Replicas`.
+
+These "MachinePool Machines" will open up the following opportunities:
+
+- Share common behavior for Machines in MachinePools, which otherwise would need to be implemented by each infrastructure provider.
+- Allow both Cluster API-driven and externally driven interactions with individual Machines in a MachinePool.
+- Enable MachinePools to use deployment strategies similarly to MachineDeployments.
+
+## Motivation
+
```

Comment:
I think of MachinePools as a choice that lets you benefit from particular provider native scalable resources capabilities for things like autoscaling, health-checking rollouts, etc. Since this is moving MachinePool closer to MachineDeployments and provider agnostic logic, can we may be add a note on benefits / reasons /differences why still to choose a MachinePool over a MachineDeployment?

Reply:
I'll tag @mboersma and @devigned to elaborate as well, but I believe the logical distinction is that MachineDeployments represent a group of individual machines/VMs on provider that are managed entirely by CAPI while MachinePools represent a provider-specific resource representing a black-box "pool" of machines. In Azure, a MachinePool represents Virtual Machine Scale Sets, and AWS and GCP should have similar constructs. Maintaining this logical distinction comes in handy for providers that don't have a MachinePool-type resource like CAPD.

Reply:
> MachinePools as a choice that lets you benefit from particular provider native scalable resources capabilities for things like autoscaling, health-checking rollouts, etc.

I think of it similarly. MachinePools allow us to decide in the infra provider which features of the cloud compute pool we would like to use rather than managing all behaviors in Cluster API while still conforming to some base common behaviors. Like you said @enxebre, autoscaling, health-checking, zone balancing and many others come to mind.